pkg/apis/pravega/v1alpha1/pravega.go

```diff
@@ -65,83 +67,209 @@ const (
 	// as JAVA_OPTS. See the following file for a complete list of options:
 	// https://github.com/pravega/pravega/blob/master/config/config.properties
 	Options map[string]string `json:"options"`
 
+	// JVM Options for tuning JVM of Controller process.
+	// These typically start with '-X' or '-XX'
+	// See the following link for a complete list of options
+	// https://docs.oracle.com/javase/8/docs/technotes/tools/unix/java.html
+	JVMOptions []string `json:"jvmOptions"`
+
+	// SegmentStoreServiceAccountName configures the service account used on segment store instances.
+	// If not specified, Kubernetes will automatically assign the default service account in the namespace
+	ServiceAccountName string `json:"serviceAccountName,omitempty"`
+
+	// SegmentStoreResources specifies the request and limit of resources that segmentStore can have.
+	// SegmentStoreResources includes CPU and memory resources
+	Resources *v1.ResourceRequirements `json:"resources,omitempty"`
+
+	// Configuration to be used when external access is enabled for this component
+	ExternalAccess *ExternalAccess `json:"externalAccess,omitempty"`
+
 	// CacheVolumeClaimTemplate is the spec to describe PVC for the Pravega cache.
 	// This field is optional. If no PVC spec, stateful containers will use
 	// emptyDir as volume
 	CacheVolumeClaimTemplate *v1.PersistentVolumeClaimSpec `json:"cacheVolumeClaimTemplate"`
+}
+
+// PravegaSpec defines the configuration of Pravega
+type PravegaSpec struct {
+	// Image defines the Pravega Docker image to use.
+	// By default, "pravega/pravega" will be used.
+	Image *PravegaImageSpec `json:"image"`
+
+	// Controller confugration
+	Controller *ControllerSpec `json:"controller"`
+
+	//SegmentStore configuration
+	SegmentStore *SegmentStoreSpec `json:"segmentstore"`
 
 	// Tier2 is the configuration of Pravega's tier 2 storage. If no configuration
 	// is provided, it will assume that a PersistentVolumeClaim called "pravega-tier2"
 	// is present and it will use it as Tier 2
 	Tier2 *Tier2Spec `json:"tier2"`
+}
 
-	// ControllerServiceAccountName configures the service account used on controller instances.
-	// If not specified, Kubernetes will automatically assign the default service account in the namespace
-	ControllerServiceAccountName string `json:"controllerServiceAccountName,omitempty"`
-
-	// SegmentStoreServiceAccountName configures the service account used on segment store instances.
-	// If not specified, Kubernetes will automatically assign the default service account in the namespace
-	SegmentStoreServiceAccountName string `json:"segmentStoreServiceAccountName,omitempty"`
+func (s *PravegaSpec) withDefaults() (changed bool) {
 
-	// ControllerResources specifies the request and limit of resources that controller can have.
-	// ControllerResources includes CPU and memory resources
-	ControllerResources *v1.ResourceRequirements `json:"controllerResources,omitempty"`
+	if s.Controller == nil {
+		fmt.Println("Controller is nil")
```

Comment:
I suppose this is for debugging? Perhaps should remove it.

Reply:
Yes, got left out from debug statements. Will remove.